### PR TITLE
Demote failed servers first during reconciliation.

### PR DESCRIPTION
**Pull request by kubawi**
In Vault, we've encountered an issue where losing a voter, then a non-voter in the same redundancy zone, in short succession, would lead to the cluster becoming unhealthy (losing an active node).

In the scenario above, the following would happen before the change in this PR (write-up stolen from @banks):
- When a voter fails it is “demoted” but what this actually means is that Autopilot will first promote the follower (with a raft reconfiguration).
- At some time later once the new voter is healthy will remove the old voter and actually “demote” them to non-voter.
- But if the new voter also fails before the first one is demoted, now we have a quorum of 4 but only 2 available servers and so the leader is forced to step down as it no longer has a majority of voters.

### CODE REVIEW

**Review by banks — COMMENTED** [review]
Great job @kubawi!

This looks like a clean implementation of the approach discussed where we ensure that we demote unhealthy servers first.

This implementation also implicitly ensures that we only change by one node in a given raft config change (actually our Raft library ensures that by only supporting a single Add/Remove node call at once).

One thing I'm not clear on: can this code change add a new way to violate `min_quorum` i.e. demote more than `f` unhealthy servers _before_ promoting their replacements? I think that is calculated elsewhere and this is just applying, but it's worth checking that those calculations don't implicitly rely on promotions always happening first (e.g. they might allow a demotion that would violate min_quorum because they also include a promotion and assume that would happen first?). That would be bad!

I'm even wondering if we need to do something more sophisticated even when `min_quorum` isn't set 🤔 .

For example, just a hypothetical case where there are 10 nodes across 5 RZs and so 5 voters..

|  RZ1 | RZ2  | RZ3  | RZ4 | RZ5 |
| --- |  --- |  --- | --- | --- | 
| A* | B* | C* | D* | E* |
| F | G | H | I | J |

If two zones went down at the same time (say 4 and 5) we could have D* and E* being demoted at the same time.

If we just demote them both first and then promote two others (say F and G), there is a point after the demotions are done where there are only 3 voters which has changed the quorum from 3 of 5 to 2 of 3 which is probably not what we want! I'm not sure if `min_quorum` being set to 3 in this case would prevent that but it should! But even assuming that `min_quorum` isn't set it would be better in this case if we instead did:

 1. Demote E
 2. Promote F
 3. Demote D
 4. Promote G

That way there would always be at least 4 voters in the configuration so we wouldn't change the quorum size.

The question is... would that re-introduce the same problem this is trying to fix if the health status of nodes is stale? In other words, what if the leader still sees I and J as healthy even though those zones have died?

Well then it would try:

 1. Demote E : OK, committed by {A, B, C}, voters now {A, B, C, D}
 3. Promote J: OK, committed by {A, B, C}, voters now {A, B, C, D, J} (but D and J are dead)
 4. Demote D: OK committed by {A, B, C}, voters now {A, B, C, J}
 5. Promote I: OK commited by {A, B, C}, voters now {A, B, C, I, J}

The cluster is still OK at this point since we still have 3 out of 5 voters healthy even though we just accidentally promoted two unhealthy servers.

Pretty soon we'd notice they are unhealthy and on a future round (say in 10 seconds) we'd demote I and J and promote two healthy standbys from other zones instead.

So I _think_ this is safe to interleave the demotions and promotions but has the benefit that it will not reduce the actual quorum size at any point.

... although that's assuming that we started off with an odd number of voters, which should be the case in general because it doesn't make sense to deploy even numbers of voters with a majority quorum scheme. But it's possible we are in an interim state where there are an even number of voters so we should consider that case. This could happen because of misconfiguration, because we started with 5 zones and one died earlier and isn't back up yet etc.

Let's say our cluster was meant to be over 5 zones, but one failed and we've been operating for a while in the following state (*means voter). Lets also assume we have a bunch of other servers unheathy (strikethrough).

|  RZ1 | RZ2  | RZ3  | RZ4 | ~RZ5~ |
| --- |  --- |  --- | --- | --- | 
| A* | B* | C* | D* | ~E~ |
| F* | ~G~ | ~H~ | ~I~ | ~J~ |

Now the same issue that has taken out G, H and I strikes F which becomes unhealthy. Next autopilot run we demote the unhealthy F but have no healthy candidates to promote (this is true both before and after this PR change).

The cluster is still up with 4 out of 5 voters! We could even tolerate one more failing. Current state:

|  RZ1 | RZ2  | RZ3  | RZ4 | ~RZ5~ |
| --- |  --- |  --- | --- | --- | 
| A* | B* | C* | D* | ~E~ |
|  ~F~ | ~G~ | ~H~ | ~I~ | ~J~ |

Now one of our standby nodes restarts and recovers and send a heartbeat to the leader just before the next AP reconcile. Sadly, we also loose RZ4 at exactly the same moment so now D is unhealthy. AP now sees:

|  RZ1 | RZ2  | RZ3  | RZ4 | ~RZ5~ |
| --- |  --- |  --- | --- | --- | 
| A* | B* | C* | ~D*~ | ~E~ |
| ~F~ | G | ~H~ | ~I~ | ~J~ |

Before this PR, it would promote G first maintaining 4 voters, before demoting D.

Now (even if we did the alternating thing I suggested above) we would demote D taking the cluster down to only 3 voters and so shrinking the quorum size to 2. before promoting G which brings it back up. 🤔 

I think we need to ensure that we can't violate `min_quorum` with this change - that seems to be the most important point. It's not obvious to me yet if we can or not.

The second part is _even if `min_quorum` isn't set_ can we still try to avoid reducing quorum size when demoting unhealthy servers by only demoting one unhealthy server at a time and only if there is an odd number of voters to start with?

I think that would be the most conservative approach:

 * Rather than demoting _all_ unhealthy first, demote at most one unhealthy voter each round before promotions (not sure if it's OK to promote more than one here though need to think)
 * Don't demote before promote if there is an even number of voters.

That would avoid introducing any new case where quorum size is changed while still solving the known issue of an RZ failure causing outage.

File: reconcile.go [hunk 1]
```diff
@@ -140,6 +155,11 @@ func (a *Autopilot) applyDemotions(state *State, changes RaftChanges) (bool, err
 			continue
 		}
 
+		if srv.Health.Healthy == demoteFailed {
+			a.logger.Debug("Ignoring demotion of server due to health status", "id", change, "healthy", srv.Health.Healthy)
```
[the remark this answers is not in the document]
Comment (kubawi) [reply reconcile.go:159]: Good point, I will update this log line when addressing the feedback above. Thanks!

[the remark this answers is not in the document]
Comment (kubawi) [reply reconcile.go]: Fixed in 57c77586c8c3c6816a16af609a678c04c6b9aa83.

File: reconcile.go [hunk 2]
```diff
@@ -32,18 +32,30 @@ func (a *Autopilot) reconcile() error {
...
+
 	// apply the promotions, if we did apply any then stop here
 	// as we do not want to apply the demotions at the same time
 	// as a means of preventing cluster instability.
 	if done, err := a.applyPromotions(state, changes); done {
 		return err
 	}
 
-	// apply the demotions, if we did apply any then stop here
+	// apply the demotions to healthy servers, if we did apply any then stop here
```
Comment (banks) [inline reconcile.go:54]: 🤔 Hmmm. What happens to unhealthy servers when they don't get handled above now? For example if there is an even number of voters we won't demote _any_ unhealthy servers at all until we've promoted another and got back to an odd number of voters. It's not immediately obvious to me if that is the correct behaviour although it might be - even number of voters is not usually a target state.

Even if we have odd voters though, we will only process one unhealthy demotion per reconcile round. I _guess_ this is OK in the sense that if that works, next reconcile round we can demote one more etc...

**Review by banks — COMMENTED** [review]
Looking great Kuba. A couple of thinkers inline. Not sure if they need changes yet but curious to hear what you and others think.

File: reconcile_test.go [hunk 3]
```diff
@@ -330,6 +352,226 @@ func TestReconcile(t *testing.T) {
...
+				Promotions: []raft.ServerID{
+					// both promotions should happen
+					"1f07052e-53c7-4f99-9cb6-6bb5b39ce8a8",
+					"bcd603a7-18e2-48c6-ac60-167e1556f4b0",
+				},
+				Demotions: []raft.ServerID{
+					// no demotions should happen in this reconciliation
+					// since we are starting with an even number of voters
+					"0a79bbf7-7113-4947-a257-6179326f188c",
+					"b8508007-68d5-42c9-92a6-28686676867e",
```
[the remark this answers is not in the document]
Comment (kubawi) [reply reconcile_test.go:557]: The changes in this test are an input to the reconcile function, not the expected outcome we assert against. The assertions are the mock functions being called. (I am not sure if you knew that or not, but this test wasn't the easiest to parse for me, so I am clarifying this just in case 😄.)

I now see that this test case is confusing, since that set of changes would possibly never be passed to `reconcile()` in reality, but I added these to assure that the function logic ignores any demotions when we have an even number of votes.

I think the overall test will be a bit easier to reason about once I implement the changes you suggested in the comment above (simpler UUIDs and helper functions for creating state). I will also add more comments to make the intent clearer.

File: reconcile_test.go [hunk 4]
```diff
@@ -330,6 +352,226 @@ func TestReconcile(t *testing.T) {
...
+						State:  RaftNonVoter,
+						Health: ServerHealth{Healthy: true},
+					},
+				},
+				Voters: []raft.ServerID{
+					"96be11f3-c9b9-45ab-a719-dc9472ada6fe",
+					"4b92b892-ee0d-4644-84fb-3117448a0401",
+					"0a79bbf7-7113-4947-a257-6179326f188c",
+					"b8508007-68d5-42c9-92a6-28686676867e",
+					"1f07052e-53c7-4f99-9cb6-6bb5b39ce8a8",
```
[the remark this answers is not in the document]
Comment (kubawi) [reply reconcile_test.go:437]: Great suggestions, thanks Paul. I will implement these in the next push.

File: reconcile.go [hunk 5]
```diff
@@ -32,18 +32,30 @@ func (a *Autopilot) reconcile() error {
...
+	// Apply the demotions to the failed servers first, but only once per a
+	// reconciliation round and if the number of voters is odd. In that case we
+	// don't want to start with the promotions, since that temporalily inflates
+	// quorum, which could lead to cluster failure. If we did apply any, then stop
+	// here as we do not want to apply the promotions at the same time as a means of
+	// preventing cluster instability.
+	if len(state.Voters)%2 != 0 {
+		if done, err := a.applyDemotions(state, changes, true); done {
+			return err
+		}
```
[the remark this answers is not in the document]
Comment (kubawi) [reply reconcile.go:44]: I agree. The reason I made the logic stop the reconcile round after a single demotion, was because I was trying to adhere to the current practice in this part of code. Having thought about it some more now, I also do not see a reason to wait for the next reconcile loop to promote healthy nodes. 

I will update the code for now and we can always change it again if we find a valid reason to keep demotion of failed node and promotions in the same round.

File: reconcile.go (hunk 2 above)
Comment (kubawi) [reply reconcile.go:54]: Sorry, this comment took a bit longer than the others because I needed to think about it and I ended up testing it empirically to make sure I didn't misunderstand the logic. 

TLDR is that it should be fine, since AP promotes, then demotes over two reconcile rounds, taking care of the unhealthy voter. Here is a step-by-step that I wrote up for myself, but I think is worth sharing.

**t0 (starting point):** 
- 4 redundancy zones
- 8 nodes
- 4 voters
- quorum of 3

| RZ1 | RZ2 | RZ3 | RZ4 |
| --- | --- | --- | --- |
| A*  | B*  | C*  | D*  |
| E   | F   | G   | H   |

**t1: `D` becomes unhealthy**

| RZ1 | RZ2 | RZ3 | RZ4    |
| --- | --- | --- | ------ |
| A*  | B*  | C*  | ~~D*~~ |
| E   | F   | G   | H      |

**t2 (reconcile loop 1):** 
- AP skips demoting `D`, because we still have an even number of voters (four).
- AP promotes `H`, we have 5 voters, quorum is still 3.

| RZ1 | RZ2 | RZ3 | RZ4    |
| --- | --- | --- | ------ |
| A*  | B*  | C*  | ~~D*~~ |
| E   | F   | G   | H*     |

**t3 (reconcile loop 2):**
- AP demotes `D` because it's unhealthy and we have an odd number of voters going into the reconcile round (five).
- We now have 4 voters, quorum was stable the whole time at 3.

| RZ1 | RZ2 | RZ3 | RZ4   |
| --- | --- | --- | ----- |
| A*  | B*  | C*  | ~~D~~ |
| E   | F   | G   | H*    |

One thing I am noticing here, is that we could possibly demote an unhealthy voter after the promotions in the same loop to avoid waiting the extra 10 seconds 🤔 I would welcome any opinions here. 

> Even if we have odd voters though, we will only process one unhealthy demotion per reconcile round. I guess this is OK in the sense that if that works, next reconcile round we can demote one more etc...

That is true. I guess we could look into demoting more than one at a time, but then we would need to add some logic to guard the `min_quorum` value, which currently (as in with the proposed changes) is guaranteed to be maintained because we only demote one voter and only when we have an odd number of them. On one hand, this seems like more complexity, on the other maybe it's more explicit. I am not sure if that optimisation is worth complicating the code more.

**Review by mkeeler — CHANGES_REQUESTED** [review]
I am pretty sure this change will break a few things:

* enterprise auto-upgrades
* unassisted replacement style CE upgrades
* RZ reconciliation when there are not enough healthy voters to promote.

The core issue is that its possible for autopilot to remain in a state where it wants to demote an unhealthy voter but where there are an even number of nodes. The state machine will be stuck until one of the following occurs:

* An out-of-band change is made to the Raft configuration to alter suffrage.
* Other servers are added that might need promotion and then having the promotions applied results in a cluster with an odd number of voters.

For auto-upgrades, its pretty much guaranteed during normal healthy operation to temporarily be in a state with an even number of voters. If at that point in time you want to demote one unhealthy server on the version being upgraded from, it would not happen and the upgrade would be stuck indefinitely.

For vanilla CE replacement style upgrades we are in a similar boat. If an old server becomes unhealthy as new servers are added you can end up with having an even number of voters, nothing left to promote but being unable to demote anything failed. 

Consider the case where 3 new servers are added to an existing cluster of 3 voters with one being unhealthy. The suffrage would transition like so:

2/1 (old) -> 2/1 (old) + 3/0 (new) -> 0/1 (old) + 3/0 (new)

At that point we still have an even number of nodes, we still need to demote the old server (its never coming back).

I _think_ the right solution then is to change the `applyDemotions` func to operate in one of two modes:

1. Demote a single unhealthy voter (called when num voters is odd prior to any promotions)
2. Demote everything as desired. (called only when there were no promotions to be done)

That is in contrast to the code as implemented which will operate in one of these two modes:

1. Demote a single unhealthy voter (called when num voters is odd prior to any promotions)
2. Demote all healthy servers as desired. (called only when there were no promotions to be done).

**Comment by banks** [comment]
@mkeeler yeah I also picked up on the split between unhealthy and healthy as well as the 1 limit leading to asymmetric situations. Previously all the examples I'd thought up Kuba thought through and we found they would eventually work out OK as far as we can tell.

It sounds like you've thought of others that wouldn't although I've not totally understood them yet.

But in general, it seems much safer and easier to reason about with the change that you propose. To state it a different way:
 * When the number of voters is odd, use the new behaviour where at-most one single unhealthy server is demoted first, then promotions...
 * When the number of voters is even, **use the same logic as before this PR**, i.e. promote all then demote all.

Even then I still feel like the asymmetry of only demoting one unhealthy node on odd is strange. I think it's OK because it will make the cluster revert to "demote everything" on the next round (probably assuming no other cluster changes happen out of band). But it would still feel easier to reason about to me if we made this change the most minimal one it can be to solve the issue which seems to be:

 1. If Odd(numVoters): 
     * attempt to demote at most 1 unhealthy
     * attempt promotions, if we do some return
     * attempt **all** other demotions that weren't that first unhealthy one
 2. If Even(numVoters):
     * Exactly the same behaviour as before the PR (all promotions, all demotions).

@kubawi what do you think? Was there some reason that the more minimal change was incorrect or more complex to implement?

**Comment by kubawi** [comment]
@banks Embarrassingly enough, I cannot really recall why I picked this approach over a simpler one. It could have been that that's where my head went first and then I didn't find any reasons not to use this approach until Matt pointed out the stuck auto-upgrade scenario above. 

@mkeeler Thanks a lot for the review. I completely missed this scenario! 🤦 I will update the PR shortly to follow your and Paul's recommendations.

**Comment by mkeeler** [comment]
@banks I might be able to describe the issue better.

First is that with auto-upgrades will always results in a state where there are an even number of voters. The auto-upgrade flow is generally:

* idle -> await-new-voters
  * State transition happens once server with a newer version has been added.
  * This state is informational and is meant to signal to the operator that autopilot is still waiting for more new version servers to be added.
* await-new-voters -> promoting
  * State transition happens once the number of healthy potential voters on the new version is greater than or equal to the number of voters on the old version.
  * Note that whether the old version voters are healthy doesn't matter.
  * When in this state, autopilot will be promoting the new version servers (still takes into account read replicas and redundancy zones so that not all new version servers are promoted)
* promoting -> demoting
  * State transition happens once the number of voters on the new version is greater than or equal to the number of voters on the old version.
  * The raft changes the `Promoter` will ask autopilot to perform will be only demotions with no promotions.

If everything is healthy, the new code would:

* not execute demotion of unhealthy voters because the server count is even already.
* will not execute promotions because there are none being requested by the `Promoter`
* will execute demotion of healthy voters on the old version. This will demote all 3 within one call to `applyDemotions`.

Now if there is one unhealthy voter on the old version then the new code would:

* not execute demotion of unhealthy voters because the server count is even already.
* will not execute promotions because there are none being requested by the `Promoter`
* will execute demotion of healthy voters on the old version. This will demote 2 out of 3 within one call to `applyDemotions` because the remaining voter is unhealthy.

Now the cluster will be in a state with 3 healthy voters on the new version and 1 unhealthy voter on the old version. The next reconciliation round would result in:

* not executing demotions of an unhealthy server because there are an even number of servers
* not executing promotions because there are none.
* demoting all healthy servers as requests. This will not actually demote anything due to there being no servers requested for demotion that are healthy.

Then auto-upgrades will be stuck in this state until the old server becomes healthy, or another voter gets added making the voter count odd and allowing the demotion to succeed.

File: reconcile.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (kubawi) [reply reconcile.go]: Fixed in 165e61e5d149b52eb40db3756a8af0fd4f414d75 (I am cleaning up the threads on the PR, hence this comment now 😄)

**Review by mkeeler — APPROVED** [review]

File: reconcile.go [hunk 6]
```diff
@@ -140,13 +159,23 @@ func (a *Autopilot) applyDemotions(state *State, changes RaftChanges) (bool, err
...
 		a.logger.Info("Demoting server", "id", srv.Server.ID, "address", srv.Server.Address, "name", srv.Server.Name)
 
 		if err := a.demoteVoter(srv.Server.ID); err != nil {
 			return true, fmt.Errorf("failed demoting server %s: %v", srv.Server.ID, err)
 		}
 
 		demoted = true
+
+		// We only want to demote one failed server at a time to prevent violating the minimum quorum setting.
+		if demoteSingleFailedServer {
```
Comment (banks) [inline reconcile.go:176]: What if we failed to demote it above? i.e. `demoted = false` 🤔 . I'm not sure under what circumstances that could happen.

I think the answer is that we should probably still limit our selves to 1 because the most likely way we could fail to demote I can think of is that we timeout or loose connection before we find out if it worked. In that case we can't assume that `err != nil` means that it actually failed so continuing to attempt another unhealthy demotion could violate the "only one" limit.

I guess the result is that we'd return the error stop the reconcile loop and then try again in 10 seconds.

Yeah overall I agree this is the most sensible apporach, just had to think it through!

File: reconcile_test.go [hunk 7]
```diff
@@ -113,223 +66,234 @@ func TestReconcile(t *testing.T) {
...
+				m.On("DemoteVoter",
+					raft.ServerID("44444444-4444-4444-4444-444444444444"),
+					uint64(0),
+					time.Duration(0),
+				).Return(&raftIndexFuture{}).Once()
+				m.On("DemoteVoter",
+					raft.ServerID("55555555-5555-5555-5555-555555555555"),
+					uint64(0),
+					time.Duration(0),
+				).Return(&raftIndexFuture{}).Once()
```
Comment (banks) [inline reconcile_test.go:294]: It looked off to me at first to see us demoting a full half of the cluster in one round 😱 . But I guess this test is not setting a min_quorum so this is OK because it's actually doing each as a separate config change and reducing quorum as it goes, ending up leaving a cluster with 3 healthy servers in a quorum right?

**Review by banks — DISMISSED** [review]
Yes @kubawi, such awesome work.

I realised a few of these comments were from a while back but I forgot to submit them so I'm sorry for that. All optional code readability suggestions or no-op commentary of me convincing myself of things!

File: reconcile_test.go (hunk 7 above)
Comment (kubawi) [reply reconcile_test.go:294]: Yep, that was my thinking. These would all be done in a single reconcile loop (this is the behaviour from before this PR anyways), but separate raft reconfigurations. Mainly though, I wanted to ensure that we have a test case asserting on multiple demotions, regardless of health status, happening in a single reconcile round that starts with an even number of voters.

File: reconcile_test.go [hunk 8]
```diff
@@ -361,6 +282,26 @@ func TestReconcile(t *testing.T) {
 	}
 }
 
+// makeServerState creates a ServerState and returns a pointer to it. The node
+// index argument is a string that should be a single digit, used to generate
+// the server ID, name, and address.
+func makeServerState(nodeIndex string, nodeState RaftState, healthy bool) *ServerState {
```
[the remark this answers is not in the document]
Comment (kubawi) [reply reconcile_test.go]: That's totally not nit-picky, code readability should be paramount! 😁 Fixed in 413b500.

File: reconcile.go [hunk 9]
```diff
@@ -32,18 +32,28 @@ func (a *Autopilot) reconcile() error {
...
 	changes := a.promoter.CalculatePromotionsAndDemotions(conf, state)
 
+	// Apply the demotion to a failed server first, but only a single one per a
+	// reconciliation round and if the number of voters is odd. In that case we
+	// don't want to start with the promotions, since that temporalily inflates
+	// quorum, which could lead to cluster failure if more voters fail.
+	if len(state.Voters)%2 != 0 {
+		if _, err := a.applyDemotions(state, changes, true); err != nil {
+			return err
+		}
```
[the remark this answers is not in the document]
Comment (kubawi) [reply reconcile.go:42]: Yep, good point. Fixed in e7d2fc3.

**Review by banks — APPROVED** [review]
Nice!

All looking good to me. Great job Kuba!

**Review by raskchanky — DISMISSED** [review]

**Review by banks — APPROVED** [review]
Good typo fixes!
